### ISSUE TITLE
pkg: avoid `dune fmt` capturing `ocamlformat` from the PATH.

### ISSUE THREAD

**Issue by moyodiallo** [ocaml/dune#10688]
## Desired Behavior
When `ocamlformat` is not a dependency of a dune project, `dune fmt` will capture `ocamlformat` command from the PATH. 

We could read the version from `.ocamlformat` file, lock it in `dev-tools.locks/ocamlformat_dev` directory ,  build it and make it available for `dune fmt`.

The dependencies of `ocamlformat` needs to be built separately to avoid the conflict with those in `dune-project`. The `dune.lock` packages from `dune-project` are built in `_build/_private/.pkg/` so the `dev-tools.locks/ocamlformat_dev` will be built in `_build/_private/.ocamlformat/` instead.

The case when `ocamlformat` is a dependency of dune project, the old behavior will remain, `dune fmt` captures that dependency which is built in `_build/_private/.pkg/`.

## Example

We have `.ocamlformat` file. With `dune fmt`, the first time ocamlformat is solved (`dev-tools.locks/ocamlformat_dev`) , built and used after at any time only used, until `dev-tools.locks/ocamlformat_dev` is removed.
```
  $ cd project
  $ cat .ocamlformat
  profile = default
  version = 0.26.2
  $ dune pkg lock
  Solution for dune.lock:
   - <project dep>
  $ dune fmt
  Solution for dev_tools.locks/ocamlformat_dev:
  - ocamlformat.0.26.2
  - <another dep>
  $ dune fmt
  $ rm -fr dev-tools.locks/ocamlformat_dev
  $ dune fmt
  Solution for dev_tools.locks/ocamlformat_dev:
  - ocamlformat.0.26.2
  - <another dep>
```

**Comment by moyodiallo** [comment]
A PR was already opened https://github.com/ocaml/dune/pull/10647.

**Comment by rgrinberg** [comment]
A few immediate questions come to mind:

1. Which opam repository do you use to solve for ocamlformat?
2. How do users opt out of this behavior?
3. Is it possible to pin the version of ocamlformat?
4. How do you make sure to share the version of ocamlformat between different packages?
5. How does one upgrade the dependencies of a version of ocamlformat?

**Comment by rgrinberg** [comment]
By the way, when I said that an issue was necessary, I meant an issue describing the general design and implementation of your proposed solution. Usually, this is a good thing to do before starting a large amount of work because it can help catch a lot of errors early on. This has the potential to save a lot of time.

**Comment by moyodiallo** [comment]
> By the way, when I said that an issue was necessary, I meant an issue describing the general design and implementation of your proposed solution. Usually, this is a good thing to do before starting a large amount of work because it can help catch a lot of errors early on. This has the potential to save a lot of time.

Agreed, but this one was more an exploration, the design and the implementation was not clear at the start. It's was driven by the question "How can we build a dev-tool seperatly from the project dependencies". 

I could also admit a discussion around an issue ticket would be more useful instead.

**Comment by punchagan** [comment]
I think this can now be closed, since #10647

**Comment by Alizter** [comment]
I am not so sure, I've still seen users report this on more recent versions.

**Comment by shonfeder** [comment]
I'll leave this open to track the bug with the current implementation, but fixing it will be superseded by the new tool management system, as described in https://github.com/ocaml/dune/issues/16187
